<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Orchestrating Large-Scale Single-Cell Analysis with Bioconductor: Working with large data</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="badge badge-light" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="background-color: #FF4955; border-radius: 5px">
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #000">
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
            Pre-Alpha
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/large_data.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Orchestrating Large-Scale Single-Cell Analysis with Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Orchestrating Large-Scale Single-Cell Analysis with Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Orchestrating Large-Scale Single-Cell Analysis with Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 70%" class="percentage">
    70%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 70%" aria-valuenow="70" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/large_data.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="intro-sce.html">1. Introduction to Bioconductor and the SingleCellExperiment class</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="eda_qc.html">2. Exploratory data analysis and quality control</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="cell_type_annotation.html">3. Cell type annotation</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="multi-sample.html">4. Multi-sample analyses</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        5. Working with large data
        </span>
      
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="hca.html">6. Accessing data from the Human Cell Atlas (HCA)</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="multi-sample.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="hca.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="multi-sample.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Multi-sample
        </a>
        <a class="chapter-link float-end" href="hca.html" rel="next">
          Next: Accessing data from... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Working with large data</h1>
        <p>Last updated on 2024-05-02 |
        
        <a href="https://github.com/ccb-hms/osca-carpentries/edit/main/episodes/large_data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How to work with single-cell datasets that are too large to fit in
memory?</li>
<li>How to speed up single-cell analysis workflows for large
datasets?</li>
<li>How to convert between popular single-cell data formats?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Learn how to work with out-of-memory data representations such as
HDF5.</li>
<li>Learn how to speed up single-cell analysis with parallel
computation.</li>
<li>Learn how to invoke fast approximations for essential analysis
steps.</li>
<li>Learn how to convert <code>SingleCellExperiment</code> objects to
<code>SeuratObject</code>s and <code>AnnData</code> objects.</li>
</ul></div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="motivation">Motivation<a class="anchor" aria-label="anchor" href="#motivation"></a></h1>
<p>Advances in scRNA-seq technologies have increased the number of cells
that can be assayed in routine experiments. Public databases such as <a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">GEO</a> are continually
expanding with more scRNA-seq studies, while large-scale projects such
as the <a href="https://www.humancellatlas.org/" class="external-link">Human Cell Atlas</a>
are expected to generate data for billions of cells. For effective data
analysis, the computational methods need to scale with the increasing
size of scRNA-seq data sets. This section discusses how we can use
various aspects of the Bioconductor ecosystem to tune our analysis
pipelines for greater speed and efficiency.</p>
</div>
<div class="section level1">
<h1 id="out-of-memory-representations">Out of memory representations<a class="anchor" aria-label="anchor" href="#out-of-memory-representations"></a></h1>
<p>The count matrix is the central structure around which our analyses
are based. In most of the previous chapters, this has been held fully in
memory as a dense <code>matrix</code> or as a sparse
<code>dgCMatrix</code>. Howevever, in-memory representations may not be
feasible for very large data sets, especially on machines with limited
memory. For example, the 1.3 million brain cell data set from 10X
Genomics (<a href="https://doi.org/10.1038/ncomms14049" class="external-link">Zheng et al.,
2017</a>) would require over 100 GB of RAM to hold as a
<code>matrix</code> and around 30 GB as a <code>dgCMatrix</code>. This
makes it challenging to explore the data on anything less than a HPC
system.</p>
<p>The obvious solution is to use a file-backed matrix representation
where the data are held on disk and subsets are retrieved into memory as
requested. While a number of implementations of file-backed matrices are
available (e.g., <a href="https://cran.r-project.org/web/packages/bigmemory/index.html" class="external-link">bigmemory</a>,
<a href="https://bioconductor.org/packages/matter" class="external-link">matter</a>), we will
be using the implementation from the <a href="https://bioconductor.org/packages/HDF5Array" class="external-link">HDF5Array</a>
package. This uses the popular HDF5 format as the underlying data store,
which provides a measure of standardization and portability across
systems. We demonstrate with a subset of 20,000 cells from the 1.3
million brain cell data set, as provided by the <a href="https://bioconductor.org/packages/TENxBrainData" class="external-link">TENxBrainData</a>
package.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">TENxBrainData</span><span class="op">)</span></span>
<span><span class="va">sce.brain</span> <span class="op">&lt;-</span> <span class="fu">TENxBrainData20k</span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="va">sce.brain</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment 
dim: 27998 20000 
metadata(0):
assays(1): counts
rownames: NULL
rowData names(2): Ensembl Symbol
colnames: NULL
colData names(4): Barcode Sequence Library Mouse
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>Examination of the <code>SingleCellExperiment</code> object indicates
that the count matrix is a <code>HDF5Matrix</code>. From a comparison of
the memory usage, it is clear that this matrix object is simply a stub
that points to the much larger HDF5 file that actually contains the
data. This avoids the need for large RAM availability during
analyses.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">counts</span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&lt;27998 x 20000&gt; HDF5Matrix object of type "integer":
             [,1]     [,2]     [,3]     [,4] ... [,19997] [,19998] [,19999]
    [1,]        0        0        0        0   .        0        0        0
    [2,]        0        0        0        0   .        0        0        0
    [3,]        0        0        0        0   .        0        0        0
    [4,]        0        0        0        0   .        0        0        0
    [5,]        0        0        0        0   .        0        0        0
     ...        .        .        .        .   .        .        .        .
[27994,]        0        0        0        0   .        0        0        0
[27995,]        0        0        0        1   .        0        2        0
[27996,]        0        0        0        0   .        0        1        0
[27997,]        0        0        0        0   .        0        0        0
[27998,]        0        0        0        0   .        0        0        0
         [,20000]
    [1,]        0
    [2,]        0
    [3,]        0
    [4,]        0
    [5,]        0
     ...        .
[27994,]        0
[27995,]        0
[27996,]        0
[27997,]        0
[27998,]        0</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">object.size</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>2496 bytes</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">file.info</span><span class="op">(</span><span class="fu">path</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">size</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 76264332</code></pre>
</div>
<p>Manipulation of the count matrix will generally result in the
creation of a <code>DelayedArray</code> object from the <a href="https://bioconductor.org/packages/DelayedArray" class="external-link">DelayedArray</a>
package. This remembers the operations to be applied to the counts and
stores them in the object, to be executed when the modified matrix
values are realized for use in calculations. The use of delayed
operations avoids the need to write the modified values to a new file at
every operation, which would unnecessarily require time-consuming disk
I/O.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu">counts</span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu">log2</span><span class="op">(</span><span class="va">tmp</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tmp</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&lt;27998 x 20000&gt; DelayedMatrix object of type "double":
             [,1]     [,2]     [,3] ... [,19999] [,20000]
    [1,]        0        0        0   .        0        0
    [2,]        0        0        0   .        0        0
    [3,]        0        0        0   .        0        0
    [4,]        0        0        0   .        0        0
    [5,]        0        0        0   .        0        0
     ...        .        .        .   .        .        .
[27994,]        0        0        0   .        0        0
[27995,]        0        0        0   .        0        0
[27996,]        0        0        0   .        0        0
[27997,]        0        0        0   .        0        0
[27998,]        0        0        0   .        0        0</code></pre>
</div>
<p>Many functions described in the previous workflows are capable of
accepting <code>HDF5Matrix</code> objects. This is powered by the
availability of common methods for all matrix representations (e.g.,
subsetting, combining, methods from <a href="https://bioconductor.org/packages/DelayedMatrixStats" class="external-link">DelayedMatrixStats</a>
as well as representation-agnostic C++ code using <a href="https://bioconductor.org/packages/beachmat" class="external-link">beachmat</a>. For
example, we compute QC metrics below with the same
<code>calculateQCMetrics()</code> function that we used in the other
workflows.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">scater</span><span class="op">)</span></span>
<span><span class="va">is.mito</span> <span class="op">&lt;-</span> <span class="fu">grepl</span><span class="op">(</span><span class="st">"^mt-"</span>, <span class="fu">rowData</span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol</span><span class="op">)</span></span>
<span><span class="va">qcstats</span> <span class="op">&lt;-</span> <span class="fu">perCellQCMetrics</span><span class="op">(</span><span class="va">sce.brain</span>, subsets <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>Mt <span class="op">=</span> <span class="va">is.mito</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qcstats</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 20000 rows and 6 columns
            sum  detected subsets_Mt_sum subsets_Mt_detected subsets_Mt_percent
      &lt;numeric&gt; &lt;numeric&gt;      &lt;numeric&gt;           &lt;numeric&gt;          &lt;numeric&gt;
1          3060      1546            123                  10            4.01961
2          3500      1694            118                  11            3.37143
3          3092      1613             58                   9            1.87581
4          4420      2050            131                  10            2.96380
5          3771      1813            100                   8            2.65182
...         ...       ...            ...                 ...                ...
19996      4431      2050            127                   9           2.866170
19997      6988      2704             60                   9           0.858615
19998      8749      2988            305                  11           3.486113
19999      3842      1711            129                   8           3.357626
20000      1775       945             26                   6           1.464789
          total
      &lt;numeric&gt;
1          3060
2          3500
3          3092
4          4420
5          3771
...         ...
19996      4431
19997      6988
19998      8749
19999      3842
20000      1775</code></pre>
</div>
<p>Needless to say, data access from file-backed representations is
slower than that from in-memory representations. The time spent
retrieving data from disk is an unavoidable cost of reducing memory
usage. Whether this is tolerable depends on the application. One example
usage pattern involves performing the heavy computing quickly with
in-memory representations on HPC systems with plentiful memory, and then
distributing file-backed counterparts to individual users for
exploration and visualization on their personal machines.</p>
</div>
<div class="section level1">
<h1 id="parallelization">Parallelization<a class="anchor" aria-label="anchor" href="#parallelization"></a></h1>
<p>Parallelization of calculations across genes or cells is an obvious
strategy for speeding up scRNA-seq analysis workflows.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.18/BiocParallel" class="external-link">BiocParallel</a></em>
package provides a common interface for parallel computing throughout
the Bioconductor ecosystem, manifesting as a <code>BPPARAM</code>
argument in compatible functions. We can also use
<code>BiocParallel</code> with more expressive functions directly
through the packageâ€™s interface.</p>
<div class="section level3">
<h3 id="basic-use">Basic use<a class="anchor" aria-label="anchor" href="#basic-use"></a></h3>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">BiocParallel</span><span class="op">)</span></span></code></pre>
</div>
<p><code>BiocParallel</code> makes it quite easy to iterate over a
vector and distribute the computation across workers using the
<code>bplapply</code> function. Basic knowledge of <code>lapply</code>
is required.</p>
<p>In this example, we find the square root of a vector of numbers in
parallel by indicating the <code>BPPARAM</code> argument in
<code>bplapply</code>.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu">MulticoreParam</span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">bplapply</span><span class="op">(</span></span>
<span>    X <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">4</span>, <span class="fl">9</span>, <span class="fl">16</span>, <span class="fl">25</span><span class="op">)</span>,</span>
<span>    FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span> <span class="fu">sqrt</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">}</span>,</span>
<span>    BPPARAM <span class="op">=</span> <span class="va">param</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]
[1] 2

[[2]]
[1] 3

[[3]]
[1] 4

[[4]]
[1] 5</code></pre>
</div>
<p><strong>Note</strong>. The number of workers is set to 1 due to
continuous testing resource limitations.</p>
<p>There exists a diverse set of parallelization backends depending on
available hardware and operating systems.</p>
<p>For example, we might use forking across two cores to parallelize the
variance calculations on a Unix system:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">MouseGastrulationData</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">WTChimeraData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">5</span>, type <span class="op">=</span> <span class="st">"processed"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">dec.mc</span> <span class="op">&lt;-</span> <span class="fu">modelGeneVar</span><span class="op">(</span><span class="va">sce</span>, BPPARAM <span class="op">=</span> <span class="fu">MulticoreParam</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dec.mc</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 29453 rows and 6 columns
                          mean       total        tech          bio     p.value
                     &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;    &lt;numeric&gt;   &lt;numeric&gt;
ENSMUSG00000051951 0.002800256 0.003504940 0.002856726  6.48214e-04 1.20918e-01
ENSMUSG00000089699 0.000000000 0.000000000 0.000000000  0.00000e+00         NaN
ENSMUSG00000102343 0.000000000 0.000000000 0.000000000  0.00000e+00         NaN
ENSMUSG00000025900 0.000794995 0.000863633 0.000811027  5.26059e-05 3.68975e-01
ENSMUSG00000025902 0.170777718 0.388633677 0.170893240  2.17740e-01 2.48097e-11
...                        ...         ...         ...          ...         ...
ENSMUSG00000095041  0.35571083  0.34572194   0.3364126  0.009309343    0.443249
ENSMUSG00000063897  0.49007956  0.41924282   0.4407846 -0.021541821    0.599512
ENSMUSG00000096730  0.00000000  0.00000000   0.0000000  0.000000000         NaN
ENSMUSG00000095742  0.00177158  0.00211619   0.0018073  0.000308881    0.189009
tomato-td           0.57257331  0.47487832   0.4971974 -0.022319130    0.591555
                           FDR
                     &lt;numeric&gt;
ENSMUSG00000051951 6.76330e-01
ENSMUSG00000089699         NaN
ENSMUSG00000102343         NaN
ENSMUSG00000025900 7.56203e-01
ENSMUSG00000025902 1.35619e-09
...                        ...
ENSMUSG00000095041    0.756203
ENSMUSG00000063897    0.756203
ENSMUSG00000096730         NaN
ENSMUSG00000095742    0.756203
tomato-td             0.756203</code></pre>
</div>
<p>Another approach would be to distribute jobs across a network of
computers, which yields the same result:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec.snow</span> <span class="op">&lt;-</span> <span class="fu">modelGeneVar</span><span class="op">(</span><span class="va">sce</span>, BPPARAM <span class="op">=</span> <span class="fu">SnowParam</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>For high-performance computing (HPC) systems with a cluster of
compute nodes, we can distribute jobs via the job scheduler using the
<code>BatchtoolsParam</code> class. The example below assumes a SLURM
cluster, though the settings can be easily configured for a particular
system (see <a href="https://bioconductor.org/packages/3.18/BiocParallel/vignettes/BiocParallel_BatchtoolsParam.pdf" class="external-link">here</a>
for details).</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 2 hours, 8 GB, 1 CPU per task, for 10 tasks.</span></span>
<span><span class="va">rs</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">7200</span>, memory <span class="op">=</span> <span class="fl">8000</span>, ncpus <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">bpp</span> <span class="op">&lt;-</span> <span class="fu">BatchtoolsParam</span><span class="op">(</span><span class="fl">10</span>, cluster <span class="op">=</span> <span class="st">"slurm"</span>, resources <span class="op">=</span> <span class="va">rs</span><span class="op">)</span></span></code></pre>
</div>
<p>Parallelization is best suited for CPU-intensive calculations where
the division of labor results in a concomitant reduction in compute
time. It is not suited for tasks that are bounded by other compute
resources, e.g., memory or file I/O (though the latter is less of an
issue on HPC systems with parallel read/write). In particular, R itself
is inherently single-core, so many of the parallelization backends
involve (i) setting up one or more separate R sessions, (ii) loading the
relevant packages and (iii) transmitting the data to that session.
Depending on the nature and size of the task, this overhead may outweigh
any benefit from parallel computing.</p>
</div>
</div>
<div class="section level1">
<h1 id="fast-approximations">Fast approximations<a class="anchor" aria-label="anchor" href="#fast-approximations"></a></h1>
<div class="section level2">
<h2 id="nearest-neighbor-searching">Nearest neighbor searching<a class="anchor" aria-label="anchor" href="#nearest-neighbor-searching"></a></h2>
<p>Identification of neighbouring cells in PC or expression space is a
common procedure that is used in many functions, e.g.,
<code>buildSNNGraph()</code>, <code>doubletCells()</code>. The default
is to favour accuracy over speed by using an exact nearest neighbour
(NN) search, implemented with the <span class="math inline">\(k\)</span>-means for <span class="math inline">\(k\)</span>-nearest neighbours algorithm. However,
for large data sets, it may be preferable to use a faster approximate
approach.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.18/BiocNeighbors" class="external-link">BiocNeighbors</a></em>
framework makes it easy to switch between search options by simply
changing the <code>BNPARAM</code> argument in compatible functions. To
demonstrate, we will use the wild-type chimera data for which we had
applied graph-based clustering using the Louvain algorithm for community
detection:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu">colLabels</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">clusterCells</span><span class="op">(</span><span class="va">sce</span>, use.dimred <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>                               BLUSPARAM <span class="op">=</span> <span class="fu">NNGraphParam</span><span class="op">(</span>cluster.fun <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>The above clusters on a nearest neighbor graph generated with an
exact neighbour search. We repeat this below using an approximate
search, implemented using the <a href="https://github.com/spotify/Annoy" class="external-link">Annoy</a> algorithm. This
involves constructing a <code>AnnoyParam</code> object to specify the
search algorithm and then passing it to the parameterization of the
<code>NNGraphParam()</code> function. The results from the exact and
approximate searches are consistent with most clusters from the former
re-appearing in the latter. This suggests that the inaccuracy from the
approximation can be largely ignored.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">BiocNeighbors</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu">clusterCells</span><span class="op">(</span><span class="va">sce</span>, use.dimred <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>                         BLUSPARAM <span class="op">=</span> <span class="fu">NNGraphParam</span><span class="op">(</span>cluster.fun <span class="op">=</span> <span class="st">"louvain"</span>,</span>
<span>                                                  BNPARAM <span class="op">=</span> <span class="fu">AnnoyParam</span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">table</span><span class="op">(</span>exact <span class="op">=</span> <span class="fu">colLabels</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, approx <span class="op">=</span> <span class="va">clusters</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     approx
exact   1   2   3   4   5   6   7   8   9  10  11  12  13  14
   1   88   0   0   0   2   0   0   0   2   0   0   0   0   0
   2    0 143   0   1   0   0   0   0   0   0   0   0   0   0
   3    0   0  75   0   2   0   0   0   0   0   0   0   0   0
   4    0   0   0 397   0   0   0   0   0   0   0   0   0   0
   5    0   0   0   0 391   0   0   0   0   1   0   0   0   0
   6    0   0   0   0   0 202   0   0   0   0   0   0   0   0
   7    0   0   0   0   0   8 245   0   0   1   0   0   0   0
   8    0   0   0   0   1   0   0  95   0   0   0   0   0   0
   9    1   0   0   0   1   0   0   0 106   0   0   0   0   0
   10   0   0   0   0   0   0   0   0   0 113   0   0   0   0
   11   0   0   0   0   0   0   0   0   0   0 153   0   0   0
   12   0   0   0   0   4   0   0   0   0  10   0 203   0   0
   13   0   0   0   0   0   0   0   0   0   0   0   0 146   0
   14   0   0   0   0   0   0   0   0   0   0   0   0   0  20</code></pre>
</div>
<p>Note that Annoy writes the NN index to disk prior to performing the
search. Thus, it may not actually be faster than the default exact
algorithm for small datasets, depending on whether the overhead of disk
write is offset by the computational complexity of the search. It is
also not difficult to find situations where the approximation
deteriorates, especially at high dimensions, though this may not have an
appreciable impact on the biological conclusions.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">y1</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">50000</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">y2</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">50000</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span><span class="va">y1</span>, <span class="va">y2</span><span class="op">)</span></span>
<span><span class="va">exact</span> <span class="op">&lt;-</span> <span class="fu">findKNN</span><span class="op">(</span><span class="va">Y</span>, k <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">approx</span> <span class="op">&lt;-</span> <span class="fu">findKNN</span><span class="op">(</span><span class="va">Y</span>, k <span class="op">=</span> <span class="fl">20</span>, BNPARAM <span class="op">=</span> <span class="fu">AnnoyParam</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">exact</span><span class="op">$</span><span class="va">index</span> <span class="op">!=</span> <span class="va">approx</span><span class="op">$</span><span class="va">index</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.561925</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="singular-value-decomposition">Singular value decomposition<a class="anchor" aria-label="anchor" href="#singular-value-decomposition"></a></h2>
<p>The singular value decomposition (SVD) underlies the PCA used
throughout our analyses, e.g., in <code>denoisePCA()</code>,
<code>fastMNN()</code>, <code>doubletCells()</code>. (Briefly, the right
singular vectors are the eigenvectors of the gene-gene covariance
matrix, where each eigenvector represents the axis of maximum remaining
variation in the PCA.) The default <code>base::svd()</code> function
performs an exact SVD that is not performant for large datasets.
Instead, we use fast approximate methods from the <em><a href="https://CRAN.R-project.org/package=irlba" class="external-link">irlba</a></em> and
<em><a href="https://CRAN.R-project.org/package=rsvd" class="external-link">rsvd</a></em>
packages, conveniently wrapped into the <em><a href="https://bioconductor.org/packages/3.18/BiocSingular" class="external-link">BiocSingular</a></em>
package for ease of use and package development. Specifically, we can
change the SVD algorithm used in any of these functions by simply
specifying an alternative value for the <code>BSPARAM</code>
argument.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">scater</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">BiocSingular</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># As the name suggests, it is random, so we need to set the seed.</span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">101000</span><span class="op">)</span></span>
<span><span class="va">r.out</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>, BSPARAM <span class="op">=</span> <span class="fu">RandomParam</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">str</span><span class="op">(</span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">r.out</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> num [1:2411, 1:20] 14.79 5.79 13.07 -32.19 -26.45 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:2411] "cell_9769" "cell_9770" "cell_9771" "cell_9772" ...
  ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...
 - attr(*, "varExplained")= num [1:20] 192.6 87 29.4 23.1 21.6 ...
 - attr(*, "percentVar")= num [1:20] 25.84 11.67 3.94 3.1 2.89 ...
 - attr(*, "rotation")= num [1:500, 1:20] -0.174 -0.173 -0.157 0.105 -0.132 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:500] "ENSMUSG00000055609" "ENSMUSG00000052217" "ENSMUSG00000069919" "ENSMUSG00000048583" ...
  .. ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">101001</span><span class="op">)</span></span>
<span><span class="va">i.out</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>, BSPARAM <span class="op">=</span> <span class="fu">IrlbaParam</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">str</span><span class="op">(</span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">i.out</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> num [1:2411, 1:20] -14.79 -5.79 -13.07 32.19 26.45 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:2411] "cell_9769" "cell_9770" "cell_9771" "cell_9772" ...
  ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...
 - attr(*, "varExplained")= num [1:20] 192.6 87 29.4 23.1 21.6 ...
 - attr(*, "percentVar")= num [1:20] 25.84 11.67 3.94 3.1 2.89 ...
 - attr(*, "rotation")= num [1:500, 1:20] 0.174 0.173 0.157 -0.105 0.132 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:500] "ENSMUSG00000055609" "ENSMUSG00000052217" "ENSMUSG00000069919" "ENSMUSG00000048583" ...
  .. ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...</code></pre>
</div>
<p>Both IRLBA and randomized SVD (RSVD) are much faster than the exact
SVD with negligible loss of accuracy. This motivates their default use
in many <em><a href="https://bioconductor.org/packages/3.18/scran" class="external-link">scran</a></em> and
<em><a href="https://bioconductor.org/packages/3.18/scater" class="external-link">scater</a></em>
functions, at the cost of requiring users to set the seed to guarantee
reproducibility. IRLBA can occasionally fail to converge and require
more iterations (passed via <code>maxit=</code> in
<code>IrlbaParam()</code>), while RSVD involves an explicit trade-off
between accuracy and speed based on its oversampling parameter
(<code>p=</code>) and number of power iterations (<code>q=</code>). We
tend to prefer IRLBA as its default behavior is more accurate, though
RSVD is much faster for file-backed matrices.</p>
</div>
</div>
<div class="section level1">
<h1 id="interoperability-with-popular-single-cell-analysis-ecosytems">Interoperability with popular single-cell analysis ecosytems<a class="anchor" aria-label="anchor" href="#interoperability-with-popular-single-cell-analysis-ecosytems"></a></h1>
<div class="section level2">
<h2 id="seurat">Seurat<a class="anchor" aria-label="anchor" href="#seurat"></a></h2>
<p><a href="https://satijalab.org/seurat" class="external-link">Seurat</a> is an R package
designed for QC, analysis, and exploration of single-cell RNA-seq data.
Seurat can be used to identify and interpret sources of heterogeneity
from single-cell transcriptomic measurements, and to integrate diverse
types of single-cell data. Seurat is developed and maintained by the <a href="https://satijalab.org/seurat/authors.html" class="external-link">Satija lab</a> and is
released under the <a href="https://opensource.org/license/mit/" class="external-link">MIT
license</a>.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">Seurat</span><span class="op">)</span></span></code></pre>
</div>
<p>Although the basic processing of single-cell data with Bioconductor
packages (described in the <a href="https://bioconductor.org/books/release/OSCA/" class="external-link">OSCA book</a>) and
with Seurat is very similar and will produce overall roughly identical
results, there is also complementary functionality with regard to cell
type annotation, dataset integration, and downstream analysis. To make
the most of both ecosystems it is therefore beneficial to be able to
easily switch between a <code>SeuratObject</code> and a
<code>SingleCellExperiment</code>. See also the Seurat <a href="https://satijalab.org/seurat/articles/conversion_vignette.html" class="external-link">conversion
vignette</a> for conversion to/from other popular single cell formats
such as the AnnData format used by <a href="https://scanpy.readthedocs.io/en/stable/" class="external-link">scanpy</a>.</p>
<p>Here, we demonstrate converting the Seurat object produced in
Seuratâ€™s <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">PBMC
tutorial</a> to a <code>SingleCellExperiment</code> for further analysis
with functionality from OSCA/Bioconductor. We therefore need to first
install the <a href="https://github.com/satijalab/seurat-data" class="external-link">SeuratData</a> package,
which is available from GitHub only.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"satijalab/seurat-data"</span><span class="op">)</span></span></code></pre>
</div>
<p>We then proceed by loading all required packages and installing the
PBMC dataset:</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">SeuratData</span><span class="op">)</span></span>
<span><span class="fu">InstallData</span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span></code></pre>
</div>
<p>We then load the dataset as an <code>SeuratObject</code> and convert
it to a <code>SingleCellExperiment</code>.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use PBMC3K from SeuratData</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">LoadData</span><span class="op">(</span>ds <span class="op">=</span> <span class="st">"pbmc3k"</span>, type <span class="op">=</span> <span class="st">"pbmc3k.final"</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">UpdateSeuratObject</span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span></span>
<span><span class="va">pbmc.sce</span> <span class="op">&lt;-</span> <span class="fu">as.SingleCellExperiment</span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc.sce</span></span></code></pre>
</div>
<p>Seurat also allows conversion from <code>SingleCellExperiment</code>
objects to Seurat objects; we demonstrate this here on the wild-type
chimera mouse gastrulation dataset.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">WTChimeraData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">5</span>, type <span class="op">=</span> <span class="st">"processed"</span><span class="op">)</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre>
</div>
<p>After some processing of the dataset, the actual conversion is
carried out with the <code>as.Seurat</code> function.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sobj</span> <span class="op">&lt;-</span> <span class="fu">as.Seurat</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu">Idents</span><span class="op">(</span><span class="va">sobj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltype.mapped"</span></span>
<span><span class="va">sobj</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="scanpy">Scanpy<a class="anchor" aria-label="anchor" href="#scanpy"></a></h2>
<p><a href="https://scanpy.readthedocs.io" class="external-link">Scanpy</a> is a scalable
toolkit for analyzing single-cell gene expression data built jointly
with <a href="https://anndata.readthedocs.io/" class="external-link">anndata</a>. It includes
preprocessing, visualization, clustering, trajectory inference and
differential expression testing. The Python-based implementation
efficiently deals with datasets of more than one million cells. Scanpy
is developed and maintained by the <a href="">Theis lab</a> and is
released under a <a href="https://github.com/scverse/scanpy/blob/master/LICENSE" class="external-link">BSD-3-Clause
license</a>. Scanpy is part of the <a href="https://scverse.org/" class="external-link">scverse</a>, a Python-based ecosystem for
single-cell omics data analysis.</p>
<p>At the core of scanpyâ€™s single-cell functionality is the
<code>anndata</code> data structure, scanpyâ€™s integrated single-cell
data container, which is conceptually very similar to Bioconductorâ€™s
<code>SingleCellExperiment</code> class.</p>
<p>Bioconductorâ€™s <em><a href="https://bioconductor.org/packages/3.18/zellkonverter" class="external-link">zellkonverter</a></em>
package provides a lightweight interface between the Bioconductor
<code>SingleCellExperiment</code> data structure and the Python
<code>AnnData</code>-based single-cell analysis environment. The idea is
to enable users and developers to easily move data between these
frameworks to construct a multi-language analysis pipeline across
R/Bioconductor and Python.</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">zellkonverter</span><span class="op">)</span></span></code></pre>
</div>
<p>The <code>readH5AD()</code> function can be used to read a
<code>SingleCellExperiment</code> from an H5AD file. Here, we use an
example H5AD file contained in the <em><a href="https://bioconductor.org/packages/3.18/zellkonverter" class="external-link">zellkonverter</a></em>
package.</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_h5ad</span> <span class="op">&lt;-</span> <span class="fu">system.file</span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"krumsiek11.h5ad"</span>,</span>
<span>                            package <span class="op">=</span> <span class="st">"zellkonverter"</span><span class="op">)</span></span>
<span><span class="fu">readH5AD</span><span class="op">(</span><span class="va">example_h5ad</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment 
dim: 11 640 
metadata(2): highlights iroot
assays(1): X
rownames(11): Gata2 Gata1 ... EgrNab Gfi1
rowData names(0):
colnames(640): 0 1 ... 158-3 159-3
colData names(1): cell_type
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>We can also write a <code>SingleCellExperiment</code> to an H5AD file
with the <code>writeH5AD()</code> function. This is demonstrated below
on the wild-type chimera mouse gastrulation dataset.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out.file</span> <span class="op">&lt;-</span> <span class="fu">tempfile</span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".h5ad"</span><span class="op">)</span></span>
<span><span class="fu">writeH5AD</span><span class="op">(</span><span class="va">sce</span>, file <span class="op">=</span> <span class="va">out.file</span><span class="op">)</span></span></code></pre>
</div>
<p>The resulting H5AD file can then be read into Python using scanpyâ€™s
<a href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.read_h5ad.html" class="external-link">read_h5ad</a>
function and then directly used in compatible Python-based analysis
frameworks.</p>
</div>
</div>
<div class="section level1">
<h1 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a></h1>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.3.3 (2024-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] zellkonverter_1.12.1         Seurat_5.0.3                
 [3] SeuratObject_5.0.1           sp_2.1-3                    
 [5] BiocSingular_1.18.0          BiocNeighbors_1.20.2        
 [7] bluster_1.12.0               scran_1.30.2                
 [9] MouseGastrulationData_1.16.0 SpatialExperiment_1.12.0    
[11] BiocParallel_1.36.0          scater_1.30.1               
[13] ggplot2_3.5.0                scuttle_1.12.0              
[15] TENxBrainData_1.22.0         HDF5Array_1.30.1            
[17] rhdf5_2.46.1                 DelayedArray_0.28.0         
[19] SparseArray_1.2.4            S4Arrays_1.2.1              
[21] abind_1.4-5                  Matrix_1.6-5                
[23] SingleCellExperiment_1.24.0  SummarizedExperiment_1.32.0 
[25] Biobase_2.62.0               GenomicRanges_1.54.1        
[27] GenomeInfoDb_1.38.8          IRanges_2.36.0              
[29] S4Vectors_0.40.2             BiocGenerics_0.48.1         
[31] MatrixGenerics_1.14.0        matrixStats_1.2.0           
[33] BiocStyle_2.30.0            

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22              splines_4.3.3                
  [3] later_1.3.2                   bitops_1.0-7                 
  [5] filelock_1.0.3                tibble_3.2.1                 
  [7] polyclip_1.10-6               basilisk.utils_1.14.1        
  [9] fastDummies_1.7.3             lifecycle_1.0.4              
 [11] edgeR_4.0.16                  globals_0.16.3               
 [13] lattice_0.22-6                MASS_7.3-60.0.1              
 [15] magrittr_2.0.3                plotly_4.10.4                
 [17] limma_3.58.1                  rmarkdown_2.26               
 [19] yaml_2.3.8                    metapod_1.10.1               
 [21] httpuv_1.6.14                 sctransform_0.4.1            
 [23] spam_2.10-0                   spatstat.sparse_3.0-3        
 [25] reticulate_1.35.0             pbapply_1.7-2                
 [27] cowplot_1.1.3                 DBI_1.2.2                    
 [29] RColorBrewer_1.1-3            zlibbioc_1.48.2              
 [31] Rtsne_0.17                    purrr_1.0.2                  
 [33] BumpyMatrix_1.10.0            RCurl_1.98-1.14              
 [35] rappdirs_0.3.3                GenomeInfoDbData_1.2.11      
 [37] ggrepel_0.9.5                 irlba_2.3.5.1                
 [39] spatstat.utils_3.0-4          listenv_0.9.1                
 [41] goftest_1.2-3                 RSpectra_0.16-1              
 [43] spatstat.random_3.2-3         dqrng_0.3.2                  
 [45] fitdistrplus_1.1-11           parallelly_1.37.1            
 [47] DelayedMatrixStats_1.24.0     leiden_0.4.3.1               
 [49] codetools_0.2-19              tidyselect_1.2.1             
 [51] ScaledMatrix_1.10.0           viridis_0.6.5                
 [53] spatstat.explore_3.2-7        BiocFileCache_2.10.1         
 [55] jsonlite_1.8.8                ellipsis_0.3.2               
 [57] progressr_0.14.0              ggridges_0.5.6               
 [59] survival_3.5-8                tools_4.3.3                  
 [61] ica_1.0-3                     Rcpp_1.0.12                  
 [63] glue_1.7.0                    gridExtra_2.3                
 [65] xfun_0.42                     dplyr_1.1.4                  
 [67] withr_3.0.0                   BiocManager_1.30.22          
 [69] fastmap_1.1.1                 basilisk_1.14.3              
 [71] rhdf5filters_1.14.1           fansi_1.0.6                  
 [73] digest_0.6.35                 rsvd_1.0.5                   
 [75] R6_2.5.1                      mime_0.12                    
 [77] colorspace_2.1-0              scattermore_1.2              
 [79] tensor_1.5                    spatstat.data_3.0-4          
 [81] RSQLite_2.3.5                 tidyr_1.3.1                  
 [83] utf8_1.2.4                    generics_0.1.3               
 [85] data.table_1.15.2             renv_1.0.7                   
 [87] htmlwidgets_1.6.4             httr_1.4.7                   
 [89] uwot_0.1.16                   pkgconfig_2.0.3              
 [91] gtable_0.3.4                  blob_1.2.4                   
 [93] lmtest_0.9-40                 XVector_0.42.0               
 [95] htmltools_0.5.7               dotCall64_1.1-1              
 [97] scales_1.3.0                  png_0.1-8                    
 [99] knitr_1.45                    reshape2_1.4.4               
[101] rjson_0.2.21                  nlme_3.1-164                 
[103] curl_5.2.1                    zoo_1.8-12                   
[105] cachem_1.0.8                  stringr_1.5.1                
[107] BiocVersion_3.18.1            KernSmooth_2.23-22           
[109] miniUI_0.1.1.1                parallel_4.3.3               
[111] vipor_0.4.7                   AnnotationDbi_1.64.1         
[113] pillar_1.9.0                  grid_4.3.3                   
[115] vctrs_0.6.5                   RANN_2.6.1                   
[117] promises_1.2.1                dbplyr_2.5.0                 
[119] beachmat_2.18.1               xtable_1.8-4                 
[121] cluster_2.1.6                 beeswarm_0.4.0               
[123] evaluate_0.23                 magick_2.8.3                 
[125] cli_3.6.2                     locfit_1.5-9.9               
[127] compiler_4.3.3                rlang_1.1.3                  
[129] crayon_1.5.2                  future.apply_1.11.1          
[131] plyr_1.8.9                    ggbeeswarm_0.7.2             
[133] stringi_1.8.3                 deldir_2.0-4                 
[135] viridisLite_0.4.2             munsell_0.5.0                
[137] Biostrings_2.70.3             lazyeval_0.2.2               
[139] spatstat.geom_3.2-9           dir.expiry_1.10.0            
[141] ExperimentHub_2.10.0          RcppHNSW_0.6.0               
[143] patchwork_1.2.0               sparseMatrixStats_1.14.0     
[145] bit64_4.0.5                   future_1.33.1                
[147] Rhdf5lib_1.24.2               KEGGREST_1.42.0              
[149] statmod_1.5.0                 shiny_1.8.0                  
[151] interactiveDisplayBase_1.40.0 AnnotationHub_3.10.0         
[153] ROCR_1.0-11                   igraph_2.0.3                 
[155] memoise_2.0.1                 bit_4.0.5                    </code></pre>
</div>
</div>
<div class="section level1">
<h1 id="further-reading">Further Reading<a class="anchor" aria-label="anchor" href="#further-reading"></a></h1>
<ul><li>OSCA book, <a href="https://bioconductor.org/books/release/OSCA.advanced/dealing-with-big-data.html" class="external-link">Chapter
14</a>: Dealing with big data</li>
<li>The <code>BiocParallel</code> <a href="https://bioconductor.org/packages/3.18/BiocParallel/vignettes/Introduction_To_BiocParallel.html" class="external-link">intro
vignette</a>.</li>
</ul></div>
<div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a></h1>
<div id="exercise-1-out-of-memory-representation" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-out-of-memory-representation" class="callout-inner">
<h3 class="callout-title">Exercise 1: Out of memory representation<a class="anchor" aria-label="anchor" href="#exercise-1-out-of-memory-representation"></a>
</h3>
<div class="callout-content">
<p>Write the counts matrix of the wild-type chimera mouse gastrulation
dataset to an HDF5 file. Create another counts matrix that reads the
data from the HDF5 file. Compare memory usage of holding the entire
matrix in memory as opposed to holding the data out of memory.</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1">Give me a hint</h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" aria-labelledby="headingHint1" data-bs-parent="#accordionHint1">
<div class="accordion-body">
<p>See the <code>HDF5Array</code> function for reading from HDF5 and the
<code>writeHDF5Array</code> function for writing to HDF5 from the <em><a href="https://bioconductor.org/packages/3.18/HDF5Array" class="external-link">HDF5Array</a></em>
package.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>TODO</p>
</div>
</div>
</div>
</div>
<div id="exercise-2-parallelization" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-parallelization" class="callout-inner">
<h3 class="callout-title">Exercise 2: Parallelization<a class="anchor" aria-label="anchor" href="#exercise-2-parallelization"></a>
</h3>
<div class="callout-content">
<p>Perform a PCA analysis of the wild-type chimera mouse gastrulation
dataset using a multicore backend for parallel computation. Compare the
runtime of performing the PCA either in serial execution mode, in
multicore execution mode with 2 workers, and in multicore execution mode
with 3 workers.</p>
</div>
</div>
</div>
<div id="accordionHint2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint2" aria-expanded="false" aria-controls="collapseHint2">
  <h4 class="accordion-header" id="headingHint2">Give me a hint</h4>
</button>
<div id="collapseHint2" class="accordion-collapse collapse" aria-labelledby="headingHint2" data-bs-parent="#accordionHint2">
<div class="accordion-body">
<p>Use the function <code>system.time</code> to obtain the runtime of
each job.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>TODO</p>
</div>
</div>
</div>
</div>
<div id="exercise-3-conversion-to-seurat" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-3-conversion-to-seurat" class="callout-inner">
<h3 class="callout-title">Exercise 3: Conversion to Seurat<a class="anchor" aria-label="anchor" href="#exercise-3-conversion-to-seurat"></a>
</h3>
<div class="callout-content">
<p>The <a href="https://bioconductor.org/packages/scRNAseq" class="external-link">scRNAseq</a>
package provides gene-level counts for a collection of public scRNA-seq
datasets, stored as <code>SingleCellExperiment</code> objects with
annotated cell- and gene-level metadata. Consult the vignette of the <a href="https://bioconductor.org/packages/scRNAseq" class="external-link">scRNAseq</a> package
to inspect all available datasets and select a dataset of your choice.
Convert the chosen dataset to a Seurat object and produce a PCA plot
with cells colored by a cell metadata column of your choice.</p>
</div>
</div>
</div>
<div id="accordionHint3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint3" aria-expanded="false" aria-controls="collapseHint3">
  <h4 class="accordion-header" id="headingHint3">Give me a hint</h4>
</button>
<div id="collapseHint3" class="accordion-collapse collapse" aria-labelledby="headingHint3" data-bs-parent="#accordionHint3">
<div class="accordion-body">
<p>Use Seuratâ€™s <code>DimPlot</code> function.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Show me the solution</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p>Use Seuratâ€™s <code>DimPlot</code> function.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>Out-of-memory representations can be used to work with single-cell
datasets that are too large to fit in memory</li>
<li>Parallelization of calculations across genes or cells is an
effective strategy for speeding up analysis of large single-cell
datasets</li>
<li>Fast approximations for nearest neighbor search and singular value
composition can speed up essential steps of single-cell analysis with
minimal loss of accuracy</li>
<li>Converter functions between existing single-cell data formats enable
analysis workflows that leverage complementary functionality from
poplular single-cell analysis ecosystems</li>
</ul></div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h1>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>



      </div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="multi-sample.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="hca.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="multi-sample.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Multi-sample
        </a>
        <a class="chapter-link float-end" href="hca.html" rel="next">
          Next: Accessing data from... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/ccb-hms/osca-carpentries/edit/main/episodes/large_data.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/ccb-hms/osca-carpentries/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/ccb-hms/osca-carpentries/" class="external-link">Source</a></p>
				<p><a href="https://github.com/ccb-hms/osca-carpentries/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:christopher_magnano@hms.harvard.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.4" class="external-link">sandpaper (0.16.4)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.5" class="external-link">pegboard (0.7.5)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.2" class="external-link">varnish (1.0.2)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://ccb-hms.github.io/osca-carpentries/large_data.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "single-cell RNA-seq, The Carpentries, lesson, genomics, bioconductor",
  "name": "Working with large data",
  "creativeWorkStatus": "active",
  "url": "https://ccb-hms.github.io/osca-carpentries/large_data.html",
  "identifier": "https://ccb-hms.github.io/osca-carpentries/large_data.html",
  "dateCreated": "2024-01-10",
  "dateModified": "2024-05-02",
  "datePublished": "2024-05-02"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

